[core][experimental] Handle leaf node in accelerated DAG

kevin85421 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

### Problem statement

Currently, the leaf node in the DAG will be ignored after being compiled. The reason is that the DAG node only has information about its upstream nodes ([ref](https://github.com/ray-project/ray/blob/c30b6576504d91dbcc20901f095efbe103eb5f0d/python/ray/dag/dag_node.py#L275-L283)), which are the arguments of this DAG node, and not about its downstream nodes.

In addition, the ADAG API `experimental_compile()` is always called by the output DAG node. As a result, the compile process performs a DFS from the output node, and the graph only has edges from downstream nodes to upstream nodes. That's why the leaf node is not included.

```python
def test_basic(ray_start_regular):
    worker = Worker.remote(0)
    with InputNode() as i:
        input_data = worker.read_input.bind(i)
        branch1 = worker.fwd.bind(input_data)
        branch2 = worker.bwd.bind(input_data) # leaf node
        dag = MultiOutputNode([branch1])

    compiled_dag = dag.experimental_compile()
    ref = compiled_dag.execute(10)
    print(ray.get(ref)) # [10]
    print(ray.get(worker.pop_trace.remote())) # [('FWD', 0)] -> the leaf node `bwd` isn't executed.

    compiled_dag.teardown()
```

Take the above code as an example, the following figure shows a graph that only has the edges from downstream nodes to upstream nodes, and performs a DFS from the output node (the "driver" in the bottom).

![image](https://github.com/user-attachments/assets/3c761c72-8772-42e7-a709-3d131bdf5d27)


### Proposed solution

* Step 1: Each DAG node not only has the information about its upstream nodes but also downstream nodes.
* Step 2: Find the `InputNode`
* Step 3: Perform BFS from the `InputNode`.

## Related issue number

Closes #46528

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

kevin85421 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -847,7 +847,7 @@ def _preprocess(self) -> None:
             ):
                 continue
-            if len(task.downstream_node_idxs) == 0:
+            if len(task.downstream_node_idxs) == 0 and task.dag_node.is_output_node:
```
The function preprocess is a bit tricky. If we don't explicitly specify `MultiOutputNode`, this function will create a new `MultiOutputNode` with the DAG node and `output_task_idx`, and then call preprocess again. As a result, in the second round of `preprocess`, the `len(task.downstream_node_idxs)` of the node with `is_output_node` will change from 0 to 1 because it now has a new downstream node, `MultiOutputNode`.

rkooo567 replied:
can we just check when we first compile, if the current node is not multi output node, we append it automtaically and just run preproess once?

kevin85421 replied:
Updated: https://github.com/ray-project/ray/pull/47261/commits/1a186858778b7410cd87ebfb3477ca73c1120404

kevin85421 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -704,7 +704,7 @@ def _preprocess(self) -> None:
             MultiOutputNode,
         )
 
-        self.input_task_idx, self.output_task_idx = None, None
```
If we reset `output_task_idx` here, the assertion `assert self.output_task_idx is not None` will fail.

kevin85421 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
             )
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
```
maybe we should consider whether it is possible to use BFS instead of `apply_recursive` for better readability.

rkooo567 replied:
let's do it in a separate PR!

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +58,30 @@ def __init__(
         )
+
+        # The list of nodes that use this DAG node as input.
+        self._downstream_nodes: List[DAGNode] = []
```
will there be any circular dependency in this case? maybe we should make it contain a weakref

kevin85421 replied:
It works on CI and my devbox. 

> maybe we should make it contain a weakref

Do you mean use `List["DAGNode"]` instead?

kevin85421 replied:
Updated https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea

rkooo567 replied:
yeah iiuc downstream node has a reference to upstream node and vice versa. So we should make one of the contain the weakref otherwise it will have circulare ref 

For exmaple, when you do
```
            upstream_node._downstream_nodes.append(self)

```

it could be instead

```
            upstream_node._downstream_nodes.append(weakref(self))

```

(weakref is not reference counted)

kevin85421 replied:
I wrote a summary about circular reference: https://github.com/ray-project/ray/pull/47261#issuecomment-2306118753

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +58,30 @@ def __init__(
+        scanner.clear()
+        # Update upstream dependencies.
+        for upstream_node in self._upstream_nodes:
+            upstream_node._downstream_nodes.append(self)
```
```suggestion
            upstream_node._downstream_nodes.append(self)
```
consider introducing an API like update_downstream_node!

kevin85421 replied:
Updated: https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -167,6 +185,7 @@ def experimental_compile(
         if _max_buffered_results is None:
             _max_buffered_results = ctx.max_buffered_results
 
+        self.is_output_node = True
```
can we remove this with preprocess update?

kevin85421 replied:
No, in `preprocess`, it currently assumes that a DAG node with no downstream nodes is the output node. However, in this PR, both the output node and leaf nodes have no downstream nodes.

https://github.com/ray-project/ray/blob/ea8b8e186c3a6384429e0990faf85e485dbbdada/python/ray/dag/compiled_dag_node.py#L850

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
+                # in some invalid cases, some nodes may not be descendants of the
+                # root. Therefore, we also add upstream nodes to the queue so that
+                # a meaningful error message can be raised when the DAG is compiled.
+                for node in node._downstream_nodes + node._upstream_nodes:
```
nit; but list + list op copies the entire list. Can you instead use `from itertools import chain`

kevin85421 replied:
Updated https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea

rkooo567 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
+    def _find_root(self) -> "DAGNode":
+        """Return the root node of the DAG."""
+        root = self
+        while type(root).__name__ != "InputNode":
```
why not just using isinstance?

kevin85421 replied:
I copied from https://github.com/ray-project/ray/blob/master/python/ray/dag/dag_node.py#L318

rkooo567 reviewed (commented):
looks good to me! some comments regarding preprocess..

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +58,30 @@ def __init__(
+        self._downstream_nodes: List[DAGNode] = []
+        # The list of nodes that this DAG node uses as input.
+        scanner = _PyObjScanner()
+        self._upstream_nodes: List[DAGNode] = scanner.find_nodes(
```
The API doc for `find_nodes` is a bit vague: Find top-level DAGNodes.

So here it finds top level nodes or direct upstream nodes? What does top-level nodes mean?

kevin85421 replied:
Yeah, it's pretty confusing. It's meant to find the immediate upstream nodes. I will update the comment.

kevin85421 replied:
Updated https://github.com/ray-project/ray/pull/47261/commits/54df435b62b4510d731bd9df4b766f694e54d5ea

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +354,33 @@ def __call__(self, node: "DAGNode"):
+                visited.add(node)
+                # Add all unseen downstream and upstream nodes to the queue.
+                # This function should be called by the root of the DAG. However,
+                # in some invalid cases, some nodes may not be descendants of the
```
Is this valid DAG? If it's not even valid DAG, we should throw an error at DAG layer rather than aDAG (compilation) layer.

kevin85421 replied:
> Is this valid DAG?

Not sure

> If it's not even valid DAG, we should throw an error at DAG layer rather than aDAG (compilation) layer.

It makes sense to me. I think we should unify the codepath like

* Step 1: Build a DAG.
* Step 2: Validate the DAG (e.g., check whether any nodes are not descendants of InputNode).
* Step 3:
  * Option 1: Execute (DAG)
  * Option 2: Compile -> Execute (ADAG)


How about opening an issue to track the progress?

1. This PR is compatible with the current behavior, where we validate these edge cases during aDAG compilation instead of building graphs.

2. Currently, `bfs` is only used by ADAG. It doesn't affect DAG.

rkooo567 replied:
IIC replacing apply_recursive will solve the issue?

kevin85421 replied:
`apply_recursive` currently doesn't validate the case where not all DAG nodes are descendants of `InputNode`. This validation is currently handled by ADAG. For example, `a2.inc.bind(1)` is not a descendant of `InputNode`.

https://github.com/ray-project/ray/blob/6743d3d6efe16c165edbac7c1556095168bb51f1/python/ray/dag/tests/experimental/test_accelerated_dag.py#L661-L669

ruisearch42 replied:
> in some invalid cases

BTW, do we know what are those invalid cases? Leaf node cases can be handled here if we traverse from the root, right? So these are some other cases?

Would be nice if we could be more specific in the comment.

rkooo567 replied:
> apply_recursive currently doesn't validate the case where not all DAG nodes are descendants of InputNode. This validation is currently handled by ADAG. For example, a2.inc.bind(1) is not a descendant of InputNode.

ah I meant if we make regular dag use bfs instead of apply_recursive, this validation should be done the same way for dag & adag?

kevin85421 replied:
> BTW, do we know what are those invalid cases? Leaf node cases can be handled here if we traverse from the root, right? So these are some other cases?

Update in the comment


> ah I meant if we make regular dag use bfs instead of apply_recursive, this validation should be done the same way for dag & adag?

Yes

rkooo567 commented on python/ray/dag/tests/experimental/test_accelerated_dag.py:
```diff
@@ -1134,6 +1137,54 @@ def test_intra_process_channel_with_multi_readers(self, ray_start_regular):
 
 
+class TestLeafNode:
+    def test_leaf_node_one_actor(self, ray_start_regular):
```
Can you add a test which checks if this works? The code is taken from existing dag test (for checking if scanner works properly). If you plan to address apply_recursive right after this PR, we don't need this

```
def test_output_node(shared_ray_instance):
    @ray.remote
    def f(input):
        return input

    with pytest.raises(ValueError):
        with InputNode() as input_data:
            dag = MultiOutputNode(f.bind(input_data))

    with InputNode() as input_data:
        dag = MultiOutputNode([f.bind(input_data)])

    assert ray.get(dag.execute(1)) == [1]
    assert ray.get(dag.execute(2)) == [2]

    with InputNode() as input_data:
        dag = MultiOutputNode([f.bind(input_data["x"]), f.bind(input_data["y"])])

    refs = dag.execute({"x": 1, "y": 2})
    assert len(refs) == 2
    assert ray.get(refs) == [1, 2]

    with InputNode() as input_data:
        dag = MultiOutputNode(
            [f.bind(input_data["x"]), f.bind(input_data["y"]), f.bind(input_data["x"])]
        )

    refs = dag.execute({"x": 1, "y": 2})
    assert len(refs) == 3
    assert ray.get(refs) == [1, 2, 1]
```

kevin85421 replied:
what's the relationship between this test and scanner?

kevin85421 replied:
> If you plan to address apply_recursive right after this PR, we don't need this

I will not update `apply_recursive` in this PR because it is on Ray Serve's codepath.

rkooo567 replied:
just want to make sure indexing inputs works properly. I think this is handled by apply_recursive before, and we need to make sure the new bfs logic can handle the same thing. 

```
with Inputnode() as inp:
    a = inp["a"]
    b = inp["b"]
    ...

adag.execute({"a": 1, "b": 2})
```

kevin85421 replied:
I still don't fully understand the relationship between this test and scanner. Maybe we can chat tmr.

Added tests: https://github.com/ray-project/ray/pull/47261/commits/eb1914c73f56cef6899b75910a530b916ff7d0b0

The test in this PR is a bit different from the test in `test_output_node.py`.

1. ADAG doesn't support function node.
2. A shared memory channel doesn't support to be read by the same actor more than once, so I use multiple actors in this case.
3. The way to index input is different from DAG and ADAG. For example,

https://github.com/ray-project/ray/blob/650bd2672d52341c549e5a640b1f6dfa06a1869f/python/ray/dag/tests/experimental/test_accelerated_dag.py#L357-L364

rkooo567 reviewed (approved):
generally lgtm.

> Maybe we should not make it become a recursive function. It is hard to read.

should we handle this in this PR?

ruisearch42 commented:
I have another idea to address the GC issue. How about processing the nodes (e.g., collecting downstream nodes, traversing the graph) in `CompiledDAG._preprocess()`? That way we don't need to change the structure of `DAGNode` and don't have cycles in references which causes GC leaks.

Actually I feel that is also cleaner:
* DAG itself don't need to know about downstream, so we don't add unnecessary information at that layer, but only do aDAG related stuff in `CompiledDAG`. 
* That's where all our current aDAG processing happens, it's better to put functionality at a consistent place.

kevin85421 commented:
> How about processing the nodes (e.g., collecting downstream nodes, traversing the graph) in `CompiledDAG._preprocess()`?

There are two questions here:

1. If we don't inform the upstream nodes about their downstream nodes when the node is constructed, the leaf node will not be reachable in `_preprocess()`. As the PR description said, when `experimental_compile` is called, it will perform a DFS from the output node in a graph that only has the edges from downstream nodes to upstream nodes.

2. If we don't inform the upstream nodes about their downstream nodes when the node is constructed, the leaf node will be freed when we call `experimental_compile` because no one references the leaf node, and it is out of scope.


```python
def test_basic(ray_start_regular):
    worker = Worker.remote(0)
    with InputNode() as i:
        input_data = worker.read_input.bind(i)
        branch1 = worker.fwd.bind(input_data)
        branch2 = worker.bwd.bind(input_data) # leaf node
        dag = MultiOutputNode([branch1])

    compiled_dag = dag.experimental_compile()
    ref = compiled_dag.execute(10)
    print(ray.get(ref)) # [10]
    print(ray.get(worker.pop_trace.remote())) # [('FWD', 0)] -> the leaf node `bwd` isn't executed.

    compiled_dag.teardown()
```

> don't have cycles in references which causes GC leaks.

I briefly skimmed through https://devguide.python.org/internals/garbage-collector/index.html#identifying-reference-cycles. The Python garbage collector supports detecting and releasing objects involved in circular references (This seems to be the same as the behavior in Java that @ruisearch42 mentioned today.)

I think ADAG doesn't introduce a lot of GC overhead:

* The DAG graph doesn't require much memory.
* There won't be many DAGs in the same process.
* Typically, DAGs are long-running, so they should remain in memory for a long time.

In addition, I searched for "doubly linked list" on Google, and all implementations involve circular references ([example](https://realpython.com/linked-lists-python/)), so I think circular referencing is not uncommon. Therefore, I think it might be fine to keep the current implementation. What do you think? @rkooo567 @ruisearch42

kevin85421 commented:
> should we handle this in this PR?

I am trying to make `_preprocess` to be a non-recursive function. If it is too complex, maybe we can use a follow up PR instead.

kevin85421 commented on python/ray/dag/compiled_dag_node.py:
```diff
@@ -841,28 +861,6 @@ def _preprocess(self) -> None:
-            self.output_task_idx = self.dag_node_to_idx[output_node]
-            # Preprocess one more time so that we have the right output node
-            # now.
-            self._preprocess()
```
this PR also removes the recursive call of `_preprocess`.

rkooo567 commented:
> In addition, I searched for "doubly linked list" on Google, and all implementations involve circular references ([example](https://realpython.com/linked-lists-python/)), so I think circular referencing is not uncommon. Therefore, I think it might be fine to keep the current implementation. What do you think? @rkooo567 @ruisearch42

Just prefer not to make tech debt if possible (circular deps are removed by gc.collect() yes). if the fix is not as simple as just using weakref I am okay to keep it this way, but please make a comment in the code.

rkooo567 commented:
btw let's remove draft and start running tests

kevin85421 commented:
TODO:

I will open issues for:
* circular references
* graph validation: https://github.com/ray-project/ray/pull/47261#discussion_r1727543342
* update `apply_recursive` https://github.com/ray-project/ray/pull/47261#discussion_r1725567583

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +60,41 @@ def __init__(
+        """
+        Retrieve upstream nodes and update their downstream dependencies.
+
+        TODO (kevin85421): Currently, the upstream nodes and downstream nodes have
+        circular references. Therefore, it relies on the garbage collector to clean
+        them up.
```
What's the consequence of this? Does it have memory leak or any side effects?

kevin85421 replied:
It will not have memory leaks. Python's garbage collector can recycle it. 

> any side effects?

https://github.com/ray-project/ray/pull/47261#issuecomment-2306118753

ruisearch42 replied:
Yeah, I guess my question was what was the issue and the TODO exactly.

kevin85421 replied:
Updated: https://github.com/ray-project/ray/pull/47261/commits/cd35b0be0947b8f0d3e5471cedcded2c76c47640

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +60,41 @@ def __init__(
         self.cache_from_last_execute = {}
 
         self._type_hint: Optional[ChannelOutputType] = ChannelOutputType()
+        self.is_output_node = False
```
Add a comment to be clear output refers to DAG output

kevin85421 replied:
update: https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +369,39 @@ def __call__(self, node: "DAGNode"):
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
+        """Breadth-first search on the DAG."""
```
nit: Breadth-first search traversal of the DAG, and apply `fn` to all nodes.

kevin85421 replied:
update: https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +369,39 @@ def __call__(self, node: "DAGNode"):
             )
         )
 
+    def bfs(self, fn: "Callable[[DAGNode], T]"):
```
Not sure if `bfs` is still accurate if the algorithm goes to upstream nodes as well. Maybe just call it `traverse` ?

kevin85421 replied:
Update to `traverse_and_apply`: https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +369,39 @@ def __call__(self, node: "DAGNode"):
+        while not isinstance(root, InputNode):
+            if len(root._upstream_nodes) == 0:
+                raise ValueError("No InputNode found in the DAG.")
+            root = root._upstream_nodes[0]
```
assert `len(root._upstream_nodes) == 1`

kevin85421 replied:
why `len(root._upstream_nodes) == 1`?

ruisearch42 replied:
Right. I misunderstood. I think this just selects an arbitrary upstream node with index 0.

ruisearch42 reviewed (commented):
Looks good otherwise.

rkooo567 commented:
let's address @ruisearch42's comments and merge it. lmk if you end up doing preprocess refactoring

kevin85421 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +369,39 @@ def __call__(self, node: "DAGNode"):
+        """
+        from ray.dag.input_node import InputNode
+
+        root = self
```
replace root with node

kevin85421 replied:
update: https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

kevin85421 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +60,41 @@ def __init__(
             other_args_to_resolve or {}
         )
+
+        # The list of nodes that use this DAG node as input.
```
update https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

kevin85421 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +60,41 @@ def __init__(
+
+        # The list of nodes that use this DAG node as input.
+        self._downstream_nodes: List["DAGNode"] = []
+        # The list of nodes that this DAG node uses as input.
```
update https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

kevin85421 commented on python/ray/dag/dag_node.py:
```diff
@@ -58,12 +60,41 @@ def __init__(
         self._type_hint: Optional[ChannelOutputType] = ChannelOutputType()
+        self.is_output_node = False
+
+    def _prepare_upstream_nodes(self) -> List["DAGNode"]:
```
update: https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

kevin85421 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +369,39 @@ def __call__(self, node: "DAGNode"):
+        root = self
+        while not isinstance(root, InputNode):
+            if len(root._upstream_nodes) == 0:
+                raise ValueError("No InputNode found in the DAG.")
```
update: https://github.com/ray-project/ray/pull/47261/commits/cfeb6020eb5951916ce4ee9a8db9a20a1cdd04ae

kevin85421 commented:
@rkooo567 I have already addressed comments. 

> lmk if you end up doing preprocess refactoring

The refactoring has been complete: https://github.com/ray-project/ray/pull/47261#discussion_r1728250783

rkooo567 reviewed (approved):
let's merge it when tests pass!

ruisearch42 commented on python/ray/dag/dag_node.py:
```diff
@@ -335,6 +372,57 @@ def __call__(self, node: "DAGNode"):
+            if len(node._upstream_nodes) == 0:
+                raise ValueError(
+                    "No InputNode found in the DAG: when traversing upwards, "
+                    "no upstream node was found for {node}."
```
need `f"no upstream node was found for {node}."`

kevin85421 replied:
Oops, good catch

kevin85421 replied:
Updated: https://github.com/ray-project/ray/pull/47261/commits/cd35b0be0947b8f0d3e5471cedcded2c76c47640
